handler.go

```diff
@@ -0,0 +1,276 @@
+// not be called and the unmarshal error will be written as a JSON
+// response.
+//
+// If an error is returned from f, it is passed through the error mapper before
+// writing as a JSON response.
+//
+// In the third form, when no error is returned, the result is written
+// as a JSON response with status http.StatusOK.
+//
+// Handle will panic if the provided function is not in one
+// of the above forms.
+func (e ErrorMapper) Handle(f interface{}) httprouter.Handle {
+	fv := reflect.ValueOf(f)
+	pt, err := checkHandleType(fv.Type())
+	if err != nil {
+		panic(errgo.Notef(err, "bad handler function"))
+	}
+	return e.handleResult(fv.Type(), handleParams(fv, pt))
+}
+
+func checkHandleType(t reflect.Type) (*requestType, error) {
+	if t.Kind() != reflect.Func {
+		return nil, errgo.New("not a function")
+	}
+	if t.NumIn() != 3 {
+		return nil, errgo.Newf("function has %d parameters, need 3", t.NumIn())
+	}
+	if t.NumOut() > 2 {
+		return nil, errgo.Newf("function has %d result parameters, need 0, 1 or 2", t.NumOut())
+	}
+	// Second and third parameters are common to all three forms.
+	if t.In(1) != paramsType {
+		return nil, errgo.Newf("second argument is %s, need %s", t.In(1), paramsType)
+	}
+	pt, err := getRequestType(t.In(2))
+	if err != nil {
+		return nil, errgo.Notef(err, "third argument cannot be used for Unmarshal")
+	}
+	if t.NumOut() < 2 {
+		if t.In(0) != httpResponseWriterType {
+			return nil, errgo.Newf("first argument is %s, need http.ResponseWriter", t.In(0))
+		}
+	} else {
+		if t.In(0) != httpHeaderType {
+			return nil, errgo.Newf("first argument is %s, need http.Header", t.In(0))
+		}
+	}
+	if t.NumOut() > 0 {
+		if et := t.Out(t.NumOut() - 1); et != errorType {
+			return nil, errgo.Newf("final result parameter is %s, need error", et)
+		}
+	}
+	return pt, nil
+}
+
+func handleParams(fv reflect.Value, pt *requestType) func(w http.ResponseWriter, req *http.Request, p httprouter.Params) ([]reflect.Value, error) {
+	ft := fv.Type()
+	arg0IsHeader := ft.NumOut() == 2
+	argStructType := ft.In(2).Elem()
+	return func(w http.ResponseWriter, req *http.Request, p httprouter.Params) ([]reflect.Value, error) {
+		if err := req.ParseForm(); err != nil {
+			return nil, errgo.WithCausef(err, ErrUnmarshal, "cannot parse HTTP request form")
+		}
+		params := Params{
+			Request: req,
+			PathVar: p,
+		}
+		pv := reflect.New(argStructType)
+		if err := unmarshal(params, pv, pt); err != nil {
+			return nil, errgo.NoteMask(err, "cannot unmarshal parameters", errgo.Is(ErrUnmarshal))
+		}
+		var arg0 reflect.Value
+		if arg0IsHeader {
+			arg0 = reflect.ValueOf(w.Header())
+		} else {
+			arg0 = reflect.ValueOf(w)
+		}
+		return fv.Call([]reflect.Value{
+			arg0,
+			reflect.ValueOf(params),
+			pv,
+		}), nil
+	}
+}
+
+func (e ErrorMapper) handleResult(ft reflect.Type, f func(w http.ResponseWriter, req *http.Request, p httprouter.Params) ([]reflect.Value, error)) httprouter.Handle {
+	switch ft.NumOut() {
```

Comment:
I am hoping that this code should remain very stable, and we have good test coverage. If it turns out to be volatile, then I totally agree that an alternative approach may be warranted. FWIW we experimented with an alternative approach, but it turned out not as clear as this.